### PR TITLE
Add toggle to  pause/unpause auto materialization

### PR DESCRIPTION
**Pull request by salazarm**
## Summary & Motivation

https://www.notion.so/dagster/Auto-materialization-Daemon-in-the-UI-0addf469f5eb44d39c0212a770d11a95

## How I Tested These Changes
Screen Shot 2023-04-14 at 5 33 50 PM
Screen Shot 2023-04-14 at 5 33 45 PM

jest

### CODE REVIEW

File: js_modules/dagit/packages/core/src/instance/DeploymentStatusProvider.tsx [hunk 1]
```diff
@@ -27,10 +35,29 @@ export const DeploymentStatusProvider: React.FC<Props> = (props) => {
...
+    codeLocations,
+    autoMaterialization,
+  ]);
 
   return (
     <DeploymentStatusContext.Provider value={value}>{children}</DeploymentStatusContext.Provider>
   );
 };
+
+function useAutomatializationStatus() {
```
Comment (salazarm) [inline js_modules/dagit/packages/core/src/instance/DeploymentStatusProvider.tsx:51]: oops

**Review by johannkm — COMMENTED** [review]
I should have noticed this in the designs, but we shortened the name of the code artifact from Auto Materialization to Auto Materialize, so we should use that term for the toggle

File: js_modules/dagit/packages/core/src/instance/DeploymentStatusProvider.tsx [hunk 2]
```diff
@@ -27,10 +35,28 @@ export const DeploymentStatusProvider: React.FC<Props> = (props) => {
...
 };
+
+function useAutoMaterializationStatus() {
+  const {data} = useQuery<GetAutoMaterializePausedQuery, GetAutoMaterializePausedQueryVariables>(
+    AUTOMATERIALIZE_PAUSED_QUERY,
+  );
+  if (data?.instance.autoMaterializePaused) {
+    return {
+      type: 'warning' as const,
+      content: <span>Auto-materialization is paused</span>,
```
Comment (johannkm) [inline js_modules/dagit/packages/core/src/instance/DeploymentStatusProvider.tsx:58]: Auto-materializing

File: js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  it('renders daemons correctly', async () => {
+    const {findByText, queryByText} = render(
+      <MockedProvider mocks={mocks} addTypename={false}>
+        <DaemonList daemonStatuses={mockDaemons as any} />
+      </MockedProvider>,
+    );
+
+    expect(await findByText('Scheduler')).toBeInTheDocument();
+    expect(await findByText('Sensors')).toBeInTheDocument();
+    expect(await findByText('Auto-materialization')).toBeInTheDocument();
```
Comment (hellendag) [inline js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx:74]: Could use synchronous `getByText` and not need to `await`?

File: js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    };
+
+    const {findByRole} = render(
+      <MockedProvider mocks={[...mocks, setAutoMaterializePausedMock]} addTypename={false}>
+        <DaemonList daemonStatuses={mockDaemons as any} />
+      </MockedProvider>,
+    );
+
+    const switchButton = await findByRole('checkbox');
+    fireEvent.click(switchButton);
```
Comment (hellendag) [inline js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx:98]: I think `fireEvent` may be removed in later versions, so might be best to stick with `userEvent.click` here.

File: js_modules/dagit/packages/core/src/instance/DaemonList.tsx [hunk 5]
```diff
@@ -55,8 +79,65 @@ export const DaemonList: React.FC<Props> = ({daemonStatuses, showTimestampColumn
...
+                  <Checkbox
+                    format="switch"
+                    checked={!data?.instance?.autoMaterializePaused}
+                    onChange={async (e) => {
+                      const checked = e.target.checked;
+                      if (!checked) {
+                        await confirm({
+                          title: 'Pause Auto-materialization?',
+                          description:
+                            'Pausing auto-materialization will prevent new materializations trigger by an auto-materialization policy.',
```
[the remark this answers is not in the document]
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/DaemonList.tsx:99]: ah yeah should be `triggered`.

File: js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx (hunk 4 above)
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx:98]: ChatGPT wrote this test and it worked lol but yeah should probably be consistent.

File: js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+      },
+    },
+  },
+];
+
+describe('DaemonList', () => {
+  it('renders daemons correctly', async () => {
+    const {findByText, queryByText} = render(
+      <MockedProvider mocks={mocks} addTypename={false}>
+        <DaemonList daemonStatuses={mockDaemons as any} />
```
[the remark this answers is not in the document]
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/__tests__/DaemonList.test.tsx:68]: ChatGPT 😅

File: js_modules/dagit/packages/core/src/instance/DaemonList.tsx [hunk 7]
```diff
@@ -55,8 +79,65 @@ export const DaemonList: React.FC<Props> = ({daemonStatuses, showTimestampColumn
...
+          <tr>
+            <td>
+              <Box flex={{direction: 'row', justifyContent: 'space-between'}}>
+                Auto-materialization
+                {loading ? (
+                  <Spinner purpose="body-text" />
+                ) : (
+                  <Checkbox
+                    format="switch"
+                    checked={!data?.instance?.autoMaterializePaused}
```
[the remark this answers is not in the document]
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/DaemonList.tsx:92]: It starts out showing a spinner so you never see the wrong checked state.

File: js_modules/dagit/packages/core/src/instance/DaemonList.tsx [hunk 8]
```diff
@@ -55,8 +79,65 @@ export const DaemonList: React.FC<Props> = ({daemonStatuses, showTimestampColumn
...
+                          title: 'Pause Auto-materialization?',
+                          description:
+                            'Pausing auto-materialization will prevent new materializations trigger by an auto-materialization policy.',
+                        });
+                      }
+                      setAutoMaterializePaused({
+                        variables: {
+                          paused: !checked,
+                        },
+                      });
```
[the remark this answers is not in the document]
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/DaemonList.tsx:106]: Not according to the GQL definitions.

@johannkm Do we need to add @capture_error to the resolvers or is it safe not to?

File: js_modules/dagit/packages/core/src/instance/DaemonList.tsx [hunk 9]
```diff
@@ -55,8 +79,65 @@ export const DaemonList: React.FC<Props> = ({daemonStatuses, showTimestampColumn
...
+              {data?.instance.autoMaterializePaused ? (
+                <Tag intent="warning">Paused</Tag>
+              ) : (
+                <DaemonHealth daemon={assetDaemon} />
+              )}
+            </td>
+            {showTimestampColumn && (
+              <td>
+                {assetDaemon.lastHeartbeatTime ? (
+                  <Group direction="row" spacing={4}>
```
[the remark this answers is not in the document]
Comment (salazarm) [reply js_modules/dagit/packages/core/src/instance/DaemonList.tsx:122]: I copy pasted this from down below 🤷🏾

**Review by bengotow — APPROVED** [review]
Permission additions look good and this seems straightforward!

The only thing I'd do one last time before merging is visit the page and command-shift-R to make sure that all the empty states and pending / loading states work. (It seems like the `?` are all in the right places, but it's tripped me up in a few last minute PRs now :) )

**Pull request merged**